lib/gollum-lib/file.rb

```diff
@@ -171,7 +171,7 @@ def get_disk_reference
       return false if @blob.is_symlink && !FS_SUPPORT_SYMLINKS
 
       # This will try to resolve symbolic links, as well
-      pathname = Pathname.new(::File.expand_path(::File.join(@wiki.repo.path, '..', @path)))
```

Comment:
Looking at this, my guess is that the `expand_path` call here is supposed to ensure that file paths are always under the repo root, in order to counter potential directory traversal exploits: passing in something like `../../../etc/password`:

```ruby
2.6.3 :002 > File.join("/path/to/repo", "../../../etc/passwd")
 => "/path/to/repo/../../../etc/passwd" 
```

Would it be an idea to:

1. Also remove `../` from paths in `BlobEntry.normalize_dir`.
2. Call `BlobEntry.normalize_dir(@path)` here to make sure such situations cannot arise

?

Reply:
On second thought, that cannot be right, as its the result of the `File.join` that is being expanded. Nevertheless: isn't this something for which we should guard?

Reply:
Perhaps such normalization/sanitization should be performed anyway at https://github.com/gollum/gollum-lib/blob/512b9667e37f35f2661cef1afce78080f715e439/lib/gollum-lib/file.rb#L77

@bartkamphorst if you have a moment, please let us know what you think.

Reply:
@dometto ah that makes sense! I changed `BlobEntry.normalize_dir` now to also strip `.` and `..`, and realized I was previously also returning `''` for a path like `...` which is still valid :wink: So I changed it around a bit to catch those cases, and added the call here in `get_disk_reference`.

I had some unrelated tests failing locally, so let's see if it passes the CI checks.

Reply:
@dometto oh I somehow only saw your first comment in this thread, let's see what @bartkamphorst thinks.

Reply:
Sorry for wait. I like the proposed solution, and prefer it over rescuing the exception and losing the option of having directories start with a tilde. 👍 Thanks @toupeira for your work on this!